### ISSUE TITLE
Handling of crashed tasks

### ISSUE THREAD

**Issue by HansHagberg** [DeepBrainChain/deepbrainchain-release#12]
On October 18, I had a (real) training task on a 2-GPU node. This task started to show faulty behavior and eventually froze completely without exiting. Log output (from container log) stopped about 12 hours into the task. It was allowed to keep running for another 11 hours (total 23 hours). At the end, there were no PIDs on the GPUs and no memory allocation from the main process (python3). The main process was running at 100% (one CPU) at that point.

A number of observations were made during this task:

1. The faulty container could not be paused, stopped or terminated. The python3 process did not respond to "kill -9" and continued running at full CPU.

2. The dbc daemon did not detect the faulty state in any visible way.

3. There were 6 other short tasks sent to this node after the faulty task while the faulty container was still running. The dbc daemon allowed this and these tasks seem to have executed correctly.

4. System had to be rebooted (power cycled) to be restored to a normal state. After first reboot, the GPU driver was not present in the system and it had to be restored. This may indicate that the original fault was related to GPU problems.

Lessons learned

. Tasks can be buggy and cause problems that will not correct themselves automatically.
. The system may **have to** be rebooted or subjected to manual intervention.
. The dbc application will need to be able signal need for manual intervention when tasks run out of control. (mail ?)
. The dbc application needs to check that containers terminate before starting new containers.

This incident points to a likely cause for downtime (buggy tasks) and loss of availability. The dbc project has indicated that machine providers will be penalized for downtime, which is a problem in cases like this. 

One can imagine a malign competitor injecting faulty tasks to lock up a large number of nodes and sabotage the availability ranking of machine providers.

Better robustness against faulty (task) software will be needed

**Comment by brightzhang2018** [comment]
good suggestion! thank you Hans
. first step , we are gonna fix the issues that the abnormal task can't be killed; and block new task launch when the abnormal task still there.

**Comment by HansHagberg** [comment]
Update:
I just had another one of the faulty tasks that lock up and never terminate.

The main task process (python3) eventually goes zombie and the container never terminates. The zombie process uses 100% cpu and can't be killed with -9 signal.

 ps -A -ostat,ppid | grep -e '[zZ]'
root     22480 22458  0 okt22 ?        00:00:00 [dbc_task.sh]
root     23628 22480 75 okt22 ?        09:58:58 [python3] <defunct>

Terminating the parent process did not work either.

The faulty container is:
9860b2e02b09        kbobrowski/tensorflow-gpu-opencv 

on node_id=2gfpp3MAB3wyMTtKNUX3hXkAEte4HB11yzTYtLrjXvc

**Comment by HansHagberg** [comment]
Another failure mode with locked up container occurred Dec 06.
0a880b8d094a        dbctraining/tensorflow1.9.0-cuda9-gpu-py3:v1.0.0

This container executed properly but failed at the last part when result was to be uploaded and hangs indefinitely . 

`dbc_upload | publish content
dbc_upload | upload content to 49.51.49.145
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:03 --:--:--     0
curl: (28) Connection timed out after 3001 milliseconds

dbc_upload | fail to upload content to 49.51.49.145
dbc_upload | upload content to 49.51.49.192
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:03 --:--:--     0
curl: (28) Connection timed out after 3001 milliseconds

dbc_upload | fail to upload content to 49.51.49.192
dbc_upload | upload content to 122.112.243.44
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  1:26:28 --:--:--     0
`

**Comment by HansHagberg** [comment]
Looks like some error handling is needed towards the upload addresses
